BLTOUCH_HS_MODE Fix for ASSISTED_TRAMMING_WIZARD

qwewer0 opened the pull request:
### Description

Right now with `BLTOUCH_HS_MODE`, BLTouch push-pin stays deployed after each measured point, so when adjusting the bed we might accidently move our hand or something under the nozzle and hit the push-pin braking it, and even after the tramming wizard procedure is over the push-pin still stays deployed, which is obviously a bad idea.

`ASSISTED_TRAMMING_WIZARD` stow after each measurement with `BLTOUCH_HS_MODE`.
<!--

We must be able to understand your proposed change from this description. If we can't understand what the code will do from this description, the Pull Request may be closed at the maintainers' discretion. Keep in mind that the maintainer reviewing this PR may not be familiar with or have worked with the code recently, so please walk us through the concepts.

-->

### Benefits

BLTouch push-pin safety for `BLTOUCH_HS_MODE`.
<!-- What does this fix or improve? -->

### Related Issues

Fixes https://github.com/MarlinFirmware/Marlin/issues/20059
<!-- Whether this fixes a bug or fulfills a feature request, please list any related Issues here. -->

qwewer0 committed:
Update menu_tramming.cpp

thisiskeithb commented:
The purpose of `BLTOUCH_HS_MODE` is to *not* stow the pin. Perhaps a sanity check should be added instead to alert the user to disable `BLTOUCH_HS_MODE` when used with `ASSISTED_TRAMMING_WIZARD`.

qwewer0 commented:
@thisiskeithb On `ASSISTED_TRAMMING_WIZARD`, we only do one measurement (manually) at a time, so I think a stow is appropriate.

qwewer0 commented:
> We're also already raising Z (comments right before the changed code):
> 
> ```
> // Users of Tramming Wizard might have a badly misaligned bed, so raise Z by the
> // length of the deployed pin (BLTOUCH stroke < 7mm)
> ```

Ok, I still think a stow is needed for ASSISTED_TRAMMING_WIZARD with BLTOUCH_HS_MODE, but then this needs more work.

sjasonsmith commented:
> Perhaps a sanity check should be added instead to alert the user to disable BLTOUCH_HS_MODE when used with ASSISTED_TRAMMING_WIZARD.

I don't think enabling ASSISTED_TRAMMING_WIZARD should prevent somebody from using BLTOUCH_HS_MODE when they want to probe their entire bed. Please don't solve this with a sanity check.

I think stowing a BLTouch is fine any time the wizard stops for user interaction. HS_MODE is there to avoid slowdowns when doing many probes in sequence.

thisiskeithb commented:
I'd recommend increasing the existing Z raise if damaging your probe is a concern.

qwewer0 commented:
> I'd recommend increasing the existing Z raise if damaging your probe is a concern.

My concern isn't that the printer would damage the push-pin, but rather that right now it stays deployed after each measured point and when adjusting the bed we might accidently move our hand or something under the nozzle and hit the push-pin braking it, and even after the procedure is over the push-pin stays deployed, which is obviously a bad idea.

sjasonsmith commented:
@qwewer0 feel free to re-open this. I think you should update the comment above the line you modified about High-speed, though.

qwewer0 commented:
@sjasonsmith Got it.

qwewer0 committed:
Update menu_tramming.cpp

sjasonsmith commented:
Can you clarify one thing for me? I've never used G35, so I don't know the precise workflow.

Does it always perform a single probe then wait for the user? It does not probe all the corners one after the other?

qwewer0 commented:
> Can you clarify one thing for me? I've never used G35, so I don't know the precise workflow.
> 
> Does it always perform a single probe then wait for the user? It does not probe all the corners one after the other?

@sjasonsmith This fix is only for `ASSISTED_TRAMMING_WIZARD`, which is only an LCD function, and that only probes a specific point if the user selects that point then hit "measure", after that, the printer waits for the user. G35 works like it did before probing all the points in one go, without stowing on `BLTOUCH_HS_MODE`.

sjasonsmith commented:
Lol, I'm getting confused by all these different mechanisms that serve very similar purposes. Thanks for the description.

sjasonsmith commented:
Oops, it looks like I broke CI with a recent merge. Ignore this from your PR, I'll get it fixed.

qwewer0 committed:
Update menu_tramming.cpp

Use Z_CLEARANCE_DEPLOY_PROBE for more clearance and avoid unnecessary Z movement after homing.

sjasonsmith committed:
Merge remote-tracking branch 'upstream/bugfix-2.0.x' into ASSISTED_TRAMMING_WIZARD_BLTOUCH_HS_MODE

sjasonsmith commented:
Can you add a call to `probe.stow()` any time the wizard exits? Your solution probably works with a BLTouch, since it auto-retracts after each probe. Other types of probes such as a "normal" servo probe will still be left deployed.

qwewer0 commented:
> Can you add a call to `probe.stow()` any time the wizard exits? Your solution probably works with a BLTouch, since it auto-retracts after each probe. Other types of probes such as a "normal" servo probe will still be left deployed.

Then wouldn't there be a problem in between probes or `PROBE_PT_STOW` is enough for that?

FanDjango commented:
BLTouch does not "auto-retract" after a probe. Not in SW-Mode and not in normal mode. It is us that retract the probe. I would "_add a call to probe.stow() any time the wizard exits_" as @sjasonsmith says

sjasonsmith commented:
Some probes are manually or otherwise slow to deploy/stow. You should probably leave it deployed until they are completely done with it.

qwewer0 committed:
Update menu_tramming.cpp

By request, adding probe.stow(); before exiting Wizard.

sjasonsmith commented:
> BLTouch does not "auto-retract" after a probe. Not in SW-Mode and not in normal mode. It is us that retract the probe. I would "_add a call to probe.stow() any time the wizard exits_" as @sjasonsmith says. You can use `PROBE_PT_STOW` but then you won't get the `PROBE_PT_RAISE` that you probably want.

So now with `BLTOUCH_HS_MODE` **disabled**, is this going to leave the pin deployed during travel? Do we need to always use `PROBE_PT_STOW` if using a BLTOUCH, regardless of HS mode?

FanDjango commented:
Regardless of probe type or 'BLTOUCH_HS_MODE', you are responsible to issue the last call to `probe_at_point(...)` that you do, with a `PROBE_PT_STOW`, otherwise the probe (whatever kind) will not be stowed. 

Trouble is, this `PROBE_PT_STOW` means you don't get a `PROBE_PT_RAISE`, thus all coders in the past have preferred to do the `PROBE_PT__RAISE`, probe the next location etc. etc. and finally do a STOW. They do that with a call to `probe.stow()` instead of coding logic to issue the *last* call to `probe_at_point(...)` with `PROBE_PT_STOW` as they typically would then want a separate Z-RAISE to be safe. And `z_raise(...)` is actually _private_ in `probe.cpp`, adding to the problem. 

Just my 2 cents on this, sorry.

qwewer0 commented:
By my test, without `BLTOUCH_HS_MODE` in `SW-Mode` the push-pin stow after each measurement, so I have no idea...

FanDjango commented:
You are right. I just checked.

FanDjango commented:
So my above statement is true only in the case of BTLOUCH combined with HS_MODE.

These final calls to `probe.stow()` all over the place are needed because `probe_at_point(...)` doesn't do the final STOW in BLTOUCH_HS_MODE, how should it know? It relies on the callers to do a final `probe_at_point(...)` with `'PROBE_PT_STOW', which they do not do because with _other_ probes it is already stowed, so a BLTOUCH in HS_MODE would remain deployed.

FanDjango commented:
Instead of unqualified `probe.stow()` operations, you would do:
```
#if ENABLED(BLTOUCH_HS_MODE)
  probe.stow()
#ENDIF 
```
Gonna test that in G34...

qwewer0 commented:
> Instead of unqualified `probe.stow()` operations, you would do:
> 
> ```
> #IF ENABLED(BLTOUCH_HS_MODE)
>   probe.stow()
> #ENDIF 
> ```



> Other types of probes such as a "normal" servo probe will still be left deployed.

So, does `probe.stow()` needed at the end for "normal" servo probe? Forr `BLTOUCH_HS_MODE` `PROBE_PT_STOW` does the job, so that doesn't need the extra stow.

FanDjango commented:
If someone has nicely coded the **last** `probe_at_point(...)` he plans to do in his sequence **with** `PROBE_PT_STOW`, even the BLTOUCH in HS_MODE will be stowed. But most doesn't do this. Most ( = all) do lots with `PROBE_PT_RAISE` and then when they are finished, the will need the STOW in case it is BLTOUCH_HS_MODE.

FanDjango commented:
Acutally it's ok like that but all users of `probe_at_point(..)` need to know the following special information:
In case of BLTOUCH **AND** BLOUCH_HS_MODE (actually the latter implies the former), a `probe_at_point(...)` will not STOW unless you tell it to.

qwewer0 commented:
> probe_at_point

This PR "changes" the `PROBE_PT_RAISE` to `PROBE_PT_STOW` if `BLOUCH_HS_MODE` is enabled, so it stow after each `probe_at_point(..)`. Is it enough or it still needs something?

FanDjango commented:
If have tested this appoach both was (with and without BLTOUCH_HS_MODE) right now and have amended my open PR #20414, which has the same situation.

FanDjango commented:
![image](https://user-images.githubusercontent.com/51046875/101896950-13f16b80-3baa-11eb-892d-468581851358.png)

FanDjango commented:
I'm testing yours right now, gimme a moment

FanDjango commented:
Ok, looks much nicer now when the movement to and from the corners with the center is done with the pin of the probe up, even with 'BLTOUCH_HS_MODE'.

I like the TERN code - you are using `PROBE_PT_STOW` as it was intended. Now though there is less potential RAISE after the probe, but it looks like it should be enough because now the pin is raised. 

Really sweet, movement looks more professional this way. No dangling pin, very little loss of responsiveness.

I wonder if now the 
```
// In BLTOUCH HS mode, the probe travels in a deployed state.
  // Users of Tramming Wizard might have a badly misaligned bed, so raise Z by the
  // length of the deployed pin (BLTOUCH stroke < 7mm)
  do_blocking_move_to_z((Z_CLEARANCE_BETWEEN_PROBES) + TERN0(BLTOUCH_HS_MODE, 7));
```
could be reduced again?

qwewer0 commented:
> ```
> // In BLTOUCH HS mode, the probe travels in a deployed state.
>   // Users of Tramming Wizard might have a badly misaligned bed, so raise Z by the
>   // length of the deployed pin (BLTOUCH stroke < 7mm)
>   do_blocking_move_to_z((Z_CLEARANCE_BETWEEN_PROBES) + TERN0(BLTOUCH_HS_MODE, 7));
> ```
> 
> could be reduced again?

Not sure what you mean, it is now only `do_blocking_move_to_z(Z_CLEARANCE_DEPLOY_PROBE);`

FanDjango commented:
Sorry, was looking at a stale github. Needed to refresh. 

Now that you are using the `PROBE_PT_STOW` in `probe_at_point(...)`, what about the `probe.stow()` at the end of the code, it shouldn't be needed at all anymore?

qwewer0 committed:
Exit stow comment tweak

FanDjango commented:
Is that STOW in DONE still needed, now that all probe operations even in `BLTOUCH_HS_MODE` do STOW always?

qwewer0 commented:
> Is that STOW in DONE still needed, now that all probe operations even in `BLTOUCH_HS_MODE` do STOW always?
> Because I haven't put that in and it works fine.


For BLTouch it doesn't need `probe.stow()` at the end, but @sjasonsmith wrote this:

> Can you add a call to `probe.stow()` any time the wizard exits? Your solution probably works with a BLTouch, since it auto-retracts after each probe. Other types of probes such as a "normal" servo probe will still be left deployed.

FanDjango commented:
And this one:

```
const float z_probed_height = probe.probe_at_point(screws_tilt_adjust_pos[tram_index], TERN(BLTOUCH_HS_MODE, PROBE_PT_STOW, PROBE_PT_RAISE), 0, true);
```
keeps "normal" probes deployed between moves to save time. Ok, I see.

FanDjango commented:
Been playing with it and looks really good now

sjasonsmith commented on Marlin/src/lcd/menu/menu_tramming.cpp:
```diff
@@ -43,11 +43,9 @@ float z_measured[G35_PROBE_COUNT] = { 0 };
-  // length of the deployed pin (BLTOUCH stroke < 7mm)
-  do_blocking_move_to_z((Z_CLEARANCE_BETWEEN_PROBES) + TERN0(BLTOUCH_HS_MODE, 7));
-  const float z_probed_height = probe.probe_at_point(screws_tilt_adjust_pos[tram_index], PROBE_PT_RAISE, 0, true);
+  do_blocking_move_to_z(Z_CLEARANCE_DEPLOY_PROBE);
```
Why did this change to `Z_CLEARANCE_DEPLOY_PROBE` instead of `Z_CLEARANCE_BETWEEN_PROBES`?

Won't `probe_at_point` already move to deploy clearance if the probe needs to be deployed?

qwewer0 replied:
When I left it as `Z_CLEARANCE_BETWEEN_PROBES`, after homing and after probing without `ASSISTED_TRAMMING_WAIT_POSITION`, the z axis moves down to `Z_CLEARANCE_BETWEEN_PROBES` then moves to the desired tramming point, and raises to `Z_CLEARANCE_DEPLOY_PROBE`.
With it changed to `Z_CLEARANCE_DEPLOY_PROBE`, it just moves to the desired tramming point without any extra z movement.

But do let me know if it is wrong.

sjasonsmith replied:
I am trying to think of what it would do with a FIX_MOUNT_PROBE that requires manual deployment high above the bed. Those would frequently have a deploy clearance of about 50mm.

Maybe just add another `TERN` here to treat BLTouch as a special case, since it stows between each point?

`do_blocking_move_to_z(TERN(BLTOUCH, Z_CLEARANCE_DEPLOY_PROBE, Z_CLEARANCE_BETWEEN_PROBES);`

qwewer0 replied:
Ok, I can see that working. 👍

qwewer0 committed:
Update menu_tramming.cpp

qwewer0 committed:
Update menu_tramming.cpp

Missing parentheses.

sjasonsmith commented:
This looks good to me. I'll let it sit here overnight in case people have anything to add, and come back to it in the morning.

qwewer0 commented:
Might be a good idea to wait for https://github.com/MarlinFirmware/Marlin/pull/20439

FanDjango commented:
@qwewer0 Why?

qwewer0 commented:
@FanDjango Maybe `PROBE_PT_STOW_AND_RAISE` would be fitting here too
https://github.com/MarlinFirmware/Marlin/blob/d63d22b11fcc1708f58afc322be6de8621c12c1f/Marlin/src/lcd/menu/menu_tramming.cpp#L48

FanDjango commented:
Maybe it would be, but `PROBE_PT_STOW_AND_RAISE` not in #20439. It was in there by mistake, I removed it.

I am playing with `PROBE_PT_STOW_AND_RAISE` on my test system. It got in there by mistake.

It should be in a separate PR, but I am probably not going to do it. Just proposing it...

qwewer0 commented:
I just saw it, that PROBE_PT_STOW_AND_RAISE isn't in that PR. 😋
